### PR TITLE
JS Promise interop

### LINKED ISSUE #2881: Add a JS-specific interop layer for promises
Same as Java interop but for [`scala.scalajs.js.Promise`](https://www.scala-js.org/api/scalajs-library/1.0.0-M2/scala/scalajs/js/Promise.html). Methods can be added to ZIO trait and companion objects via the `*PlatformSpecific` traits in the `js` branch.

### PR DESCRIPTION
**Pull request by vasilmkd**
Resolves #2881.

In line with the Scala.js recommendations: https://www.scala-js.org/api/scalajs-library/1.0.1/scala/scalajs/js/Promise.html

@regiskuckaertz as original issue author.

### CODE REVIEW

**Comment by regiskuckaertz** [comment]
Amazing ❤️ I looked into the code on Scala JS, the `toFuture` does not seem like it is doing too much: https://github.com/scala-js/scala-js/blob/bbf3629faedb59a260d39f961bc75b31ff3821a9/library/src/main/scala/scala/scalajs/js/Thenable.scala#L54-L70

Do you have bandwidth to convert this using ZIO's own Promise?

**Comment by vasilmkd** [comment]
Thanks. Will do.

**Comment by vasilmkd** [comment]
@regiskuckaertz How about now, I think it's even more straight to the point.

**Comment by vasilmkd** [comment]
@regiskuckaertz Could you please take another look? Thanks.

File: core/js/src/main/scala/zio/ZIOPlatformSpecific.scala [hunk 1]
```diff
@@ -16,6 +16,40 @@
...
+   */
+  def toPromiseJS(implicit ev: E <:< Throwable): URIO[R, JSPromise[A]] =
+    toPromiseJSWith(ev)
+
+  /**
+   * Converts the current `ZIO` to a Scala.js promise and maps the
+   * error type with `f`.
+   */
+  def toPromiseJSWith(f: E => Throwable): URIO[R, JSPromise[A]] =
+    self.mapError(f).fold(JSPromise.reject, JSPromise.resolve[A](_))
```
Comment (regiskuckaertz) [inline core/js/src/main/scala/zio/ZIOPlatformSpecific.scala]: I suggest to take possible defects in as well:
```suggestion
    self.foldCause(c => JSPromise.reject(c.squashWith(f)), JSPromise.resolve[A](_))
```

**Comment by regiskuckaertz** [comment]
Sorry for the lack of feedback. Thanks again for doing this, there's one minor nit then it looks great to me 👍

**Comment by vasilmkd** [comment]
No problem, I committed your suggestion.

**Pull request merged**